src/goto-instrument/contracts/assigns.cpp

```diff
@@ -112,14 +113,32 @@ void assigns_clauset::add_target(const exprt &target_expr)
 
 void assigns_clauset::remove_target(const exprt &target_expr)
 {
-  targets.erase(targett(*this, targett::normalize(target_expr)));
+  write_set.erase(targett(*this, targett::normalize(target_expr)));
+}
+
+void assigns_clauset::add_freely_assignable_expr(const exprt &expr)
+{
+  auto insertion_succeeded =
+    freely_assignable_set.emplace(*this, targett::normalize(expr)).second;
+
+  if(!insertion_succeeded)
+  {
+    log.warning() << "Ignored duplicate expression '"
+                  << from_expr(ns, expr.id(), expr) << "' in assigns clause at "
+                  << expr.source_location().as_string() << messaget::eom;
+  }
+}
+
+void assigns_clauset::remove_freely_assignable_expr(const exprt &expr)
+{
+  freely_assignable_set.erase(targett(*this, targett::normalize(expr)));
 }
 
 goto_programt assigns_clauset::generate_havoc_code() const
 {
   modifiest modifies;
-  for(const auto &target : targets)
-    modifies.insert(target.address.object());
+  for(const auto &target : write_set)
+    modifies.insert(to_address_of_expr(target.address).object());
```

Comment:
But `address` is already an `address_of_exprt`

Reply:
Correct. This is unnecessary. Removed it.